### PR TITLE
feat(mandates): allow off-session payments using `payment_method_id`

### PR DESCRIPTION
**Pull request by Chethan-rao**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

When a mandate is created on connector's end, we store the connector mandate details in our payment methods table. Now in order to make recurring payment with this, we basically have to do a token based payment where set the `setup_future_usage = off_session` and we list the payment methods of a customer and make a payment with payment token. But we should also allow off-session payments using payment method id without having to list customer payment methods. For this, we accept `off_session = true` and payment method id in the request and make a recurring mandate payment.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

Api contract changes - 
Added new field in `PaymentsRequest`: 

```
"recurring_details": {
      "type": "payment_method_id/mandate_id",
      "data": "String"
},
```

## How did you test it?

1. mca creation - 

```
curl --location 'http://localhost:8080/account/merchant_1710841026/connectors' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "connector_type": "fiz_operations",
    "business_country": "US",
    "business_label": "default",
    "connector_name": "cybersource",
    "connector_account_details": {
        "auth_type": "SignatureKey",
        "api_key": "abc",
        "api_secret": "xyz",
        "key1": "key"
    },
    "test_mode": true,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "payment_experience": null,
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
    "connector_webhook_details": {
        "merchant_secret": "Xavior"
    }
}'
```

2. Setup mandate on connector's end

```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: abc' \
--data-raw '{
    "amount": 0,
    "currency": "USD",
    "confirm": true,
    "profile_id": "pro_ByhEgJFlWSahAgSqaUZl",
    "capture_method": "automatic",
    "customer_id": "mandate_customer1",
    "email": "guest@example.com",
    "setup_future_usage": "off_session",
    "customer_acceptance": {
        "acceptance_type": "online"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "03",
            "card_exp_year": "2030",
            "card_holder_name": "Test Holder",
            "card_cvc": "737"
        }
    },
    "payment_type": "setup_mandate",
    "billing": {
        "address": {
            "city": "test",
            "country": "US",
            "line1": "here",
            "line2": "there",
            "line3": "anywhere",
            "zip": "560095",
            "state": "Washington",
            "first_name": "One",
            "last_name": "Two"
        },
        "phone": {
            "number": "1234567890",
            "country_code": "+1"
        },
        "email": "guest@example.com"
    },
    "authentication_type": "no_three_ds"
}'
```

db entry - 

image

3. Recurring payment using `payment_method_id` 

```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: abc' \
--data-raw '{
    "amount": 499,
    "currency": "USD",
    "confirm": true,
    "profile_id": "pro_ByhEgJFlWSahAgSqaUZl",
    "capture_method": "automatic",
    "customer_id": "mandate_customer1",
    "email": "guest@example.com",
    "off_session": true,
    "recurring_details": {
        "type": "payment_method_id",
        "data": "pm_lmTnIO5EdCiiMgRPrV9x"
    },
    "payment_method": "card",
    "billing": {
        "address": {
            "city": "test",
            "country": "US",
            "line1": "here",
            "line2": "there",
            "line3": "anywhere",
            "zip": "560095",
            "state": "Washington",
            "first_name": "One",
            "last_name": "Two"
        },
        "phone": {
            "number": "1234567890",
            "country_code": "+1"
        },
        "email": "guest@example.com"
    },
    "authentication_type": "no_three_ds"
}'
```

Payment succeded - 
image

Also try creating mandate on hyperswitch end by providing mandate_data in setup mandate request and get the `mandate_id`. Do the recurring payment with `mandate_id`. Also try another recurring payment with below structure

```
    "recurring_details": {
        "type": "mandate_id",
        "data": "xyz"
    }
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments/helpers.rs [hunk 1]
```diff
@@ -413,75 +417,127 @@ pub async fn get_token_pm_type_mandate_details(
...
+                        let payment_method_info = state
+                            .store
+                            .find_payment_method(payment_method_id)
+                            .await
+                            .to_not_found_response(
+                                errors::ApiErrorResponse::PaymentMethodNotFound,
+                            )?;
+
+                        (
+                            request.payment_token.to_owned(),
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:475]: If recurring details are provided, then token is not required, so we can hardcode this to `None`. It would be better to add validation for this as well in the PaymentsConfirm `ValidateRequest` operation (both token and recurring details cannot be given).

```suggestion
                            request.payment_token.to_owned(),
```

File: crates/router/src/core/payments/helpers.rs [hunk 2]
```diff
@@ -413,75 +417,127 @@ pub async fn get_token_pm_type_mandate_details(
...
+                            request.payment_method,
+                            request.payment_method_type,
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:477]:
```suggestion
                            payment_method_info.payment_method,
                            payment_method_info.payment_method_type,
```

File: crates/router/src/core/payments/helpers.rs [hunk 3]
```diff
@@ -413,75 +417,127 @@ pub async fn get_token_pm_type_mandate_details(
...
+                            mandate_data,
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:478]: We should also be able to hardcode this to `None`. Another validation would be required (`mandate_data` and `recurring_details` cannot be passed together).

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -939,7 +999,8 @@ pub fn create_complete_authorize_url(
 }
 
 fn validate_recurring_mandate(req: api::MandateValidationFields) -> RouterResult<()> {
-    req.mandate_id.check_value_present("mandate_id")?;
+    req.recurring_details
+        .check_value_present("recurring_details")?;
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:1003]: we need to check `mandate_id` also right? One of the two should be given.

File: crates/router/src/core/payments/helpers.rs [hunk 5]
```diff
@@ -1949,7 +2010,8 @@ pub(crate) fn validate_payment_method_fields_present(
     utils::when(
         req.payment_method.is_some()
             && req.payment_method_data.is_none()
-            && req.payment_token.is_none(),
+            && req.payment_token.is_none()
+            && req.recurring_details.is_none(),
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:2014]: Why is this required?

File: crates/router/src/core/payments/operations/payment_complete_authorize.rs [hunk 6]
```diff
@@ -247,6 +249,14 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
                 id: profile_id.to_string(),
             })?;
 
+        let recurring_details = request
```
Comment (vspecky) [inline crates/router/src/core/payments/operations/payment_complete_authorize.rs:252]: Why are we ignoring `MandateId` here?

File: crates/router/src/core/payments.rs [hunk 7]
```diff
@@ -2263,6 +2263,7 @@ where
     pub authorizations: Vec<diesel_models::authorization::Authorization>,
     pub authentication: Option<storage::Authentication>,
     pub frm_metadata: Option<serde_json::Value>,
+    pub recurring_details: Option<String>,
```
Comment (vspecky) [inline crates/router/src/core/payments.rs:2266]: It's better if this is the enum itself. With a `String`, we won't know later on in the flow whether it's a recurring mandate or a recurring MIT payment.

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/helpers.rs (hunk 4 above)
Comment (Chethan-rao) [reply crates/router/src/core/payments/helpers.rs:1003]: `recurring_details` is of type `RecurringDetails` in which during from conversion itself, I'm mapping the mandate_id to `RecurringDetails::MandateId` or else take the request.recurring_details

```
impl From<&PaymentsRequest> for MandateValidationFields {
    fn from(req: &PaymentsRequest) -> Self {
        let recurring_details = req
            .mandate_id
            .clone()
            .map(RecurringDetails::MandateId)
            .or(req.recurring_details.clone());

        Self {
            recurring_details,
            confirm: req.confirm,
            customer_id: req
                .customer
                .as_ref()
                .map(|customer_details| &customer_details.id)
                .or(req.customer_id.as_ref())
                .map(ToOwned::to_owned),
            mandate_data: req.mandate_data.clone(),
            setup_future_usage: req.setup_future_usage,
            off_session: req.off_session,
        }
    }
}
```

File: crates/router/src/core/payments/operations/payment_complete_authorize.rs (hunk 6 above)
Comment (Chethan-rao) [reply crates/router/src/core/payments/operations/payment_complete_authorize.rs:252]: because in `PaymentsRequest`, we have both `mandate_id: Option<String>` and `recurring_details: Option<String>` where recurring_details is mapped to payment method id

File: crates/router/src/core/payments/helpers.rs (hunk 5 above)
Comment (Chethan-rao) [reply crates/router/src/core/payments/helpers.rs:2014]: Now that i'm getting the `payment_method` from `payment_method_info` itself as suggested by you, we can remove this

File: crates/router/src/core/payments/helpers.rs (hunk 3 above)
Comment (Chethan-rao) [reply crates/router/src/core/payments/helpers.rs:478]: That validation is already taken care in `validate_and_get_mandate_type`

```
impl MandateValidationFieldsExt for MandateValidationFields {
    fn validate_and_get_mandate_type(
        &self,
    ) -> errors::CustomResult<Option<MandateTransactionType>, errors::ValidationError> {
        match (&self.mandate_data, &self.recurring_details) {
            (None, None) => Ok(None),
            (Some(_), Some(_)) => Err(errors::ValidationError::InvalidValue {
                message: "Expected one out of recurring_details and mandate_data but got both"
                    .to_string(),
            })
            .into_report(),
            (_, Some(_)) => Ok(Some(MandateTransactionType::RecurringMandateTransaction)),
            (Some(_), _) => Ok(Some(MandateTransactionType::NewMandateTransaction)),
        }
    }
}
```

File: openapi/openapi_spec.json [hunk 8]
```diff
@@ -15467,6 +15499,49 @@
...
+              "data": {
+                "type": "string"
+              }
+            }
+          }
+        ],
+        "discriminator": {
+          "propertyName": "type"
+        }
+      },
```
Comment (jagan-jaya) [inline openapi/openapi_spec.json:15544]:
```suggestion
      "RecurringDetails": {
        "type": "object",
        "required": [
          "type",
          "data"
        ],
        "properties": {
          "type": {
            "type": "string",
            "enum": [
              "mandate_id",
              "payment_method_id"
            ]
          },
          "data": {
            "type": "string"
          }
        },
        "discriminator": {
          "propertyName": "type"
        }
      }
```

Comment (Chethan-rao) [reply openapi/openapi_spec.json:15544]: currently we are providing support to pass `mandate_id` inside `recurring_details`. So it has to be one of `mandate_id` or `payment_method_id`. Later we'll deprecate the `mandate_id` field in root level. 

> I hope the testing for recurring payment with mandate_id in root level and payment_token is done already
yes

**Review by jagan-jaya — APPROVED** [review]

**Review by vspecky — APPROVED** [review]
